File: frontend/src/api/analysis.ts
import api from "../lib/api"

export interface DimensionScore {
  score: number
  max: number
  notes: string[]
}

export interface KeywordMatchDimension
  extends DimensionScore {
  missing_keywords: string[]
  matched_keywords: string[]
}

export interface RedFlagsDimension
  extends DimensionScore {
  flags_found: string[]
}

export interface DimensionScores {
  parseability: DimensionScore
  keyword_match: KeywordMatchDimension
  experience_alignment: DimensionScore
  impact_quantification: DimensionScore
  formatting_length: DimensionScore
  red_flags: RedFlagsDimension
}

export interface ResumeAnalysis {
  id: string
  resume_id: string

  target_role: string
  years_of_experience: number

  overall_score: number | null
  ats_score: number | null

  dimension_scores: DimensionScores | null

  skills: string[]
  strengths: string[]
  weaknesses: string[]
  missing_skills: string[]
  recommendations: string[]
  top_3_fixes: string[]

  summary: string | null
  model_name: string | null
  created_at: string
}

interface ResumeAnalysisListResponse {
  analyses: ResumeAnalysis[]
}

export interface AnalyzeResumeRequest {
  target_role: string
  years_of_experience: number
}

export async function analyzeResume(
  resumeId: string,
  data: AnalyzeResumeRequest,
): Promise<ResumeAnalysis> {
  const response = await api.post<ResumeAnalysis>(
    `/resumes/${resumeId}/analyze`,
    data,
  )

  return response.data
}

export async function getResumeAnalyses(
  resumeId: string,
): Promise<ResumeAnalysis[]> {
  const response =
    await api.get<ResumeAnalysisListResponse>(
      `/resumes/${resumeId}/analyses`,
    )

  return response.data.analyses
}

export async function getAnalysis(
  analysisId: string,
): Promise<ResumeAnalysis> {
  const response = await api.get<ResumeAnalysis>(
    `/analyses/${analysisId}`,
  )

  return response.data
}

File: backend/app/schemas/analysis.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class DimensionScore(BaseModel):
    score: int
    max: int
    notes: list[str] = []


class KeywordMatchDimension(DimensionScore):
    missing_keywords: list[str] = []
    matched_keywords: list[str] = []


class RedFlagsDimension(DimensionScore):
    flags_found: list[str] = []


class DimensionScores(BaseModel):
    parseability: DimensionScore
    keyword_match: KeywordMatchDimension
    experience_alignment: DimensionScore
    impact_quantification: DimensionScore
    formatting_length: DimensionScore
    red_flags: RedFlagsDimension


class ResumeAnalysisCreate(BaseModel):
    target_role: str
    years_of_experience: float

    overall_score: int | None = None
    ats_score: int | None = None

    dimension_scores: DimensionScores

    skills: list[str] = []
    strengths: list[str] = []
    weaknesses: list[str] = []
    missing_skills: list[str] = []
    recommendations: list[str] = []
    top_3_fixes: list[str] = []

    summary: str | None = None
    model_name: str | None = None


class ResumeAnalysisResponse(BaseModel):
    id: UUID
    resume_id: UUID

    target_role: str
    years_of_experience: float

    overall_score: int | None
    ats_score: int | None

    dimension_scores: DimensionScores

    skills: list[str]
    strengths: list[str]
    weaknesses: list[str]
    missing_skills: list[str]
    recommendations: list[str]
    top_3_fixes: list[str]

    summary: str | None
    model_name: str | None

    created_at: datetime

    model_config = {
        "from_attributes": True,
    }


class ResumeAnalysisListResponse(BaseModel):
    analyses: list[ResumeAnalysisResponse]
